File: ap_launcher/harbor.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any

import requests


@dataclass(frozen=True)
class HarborRepo:
    repo: str
    tag: str


class HarborClient:
    """Minimal Harbor v2 API client (project-scoped)."""

    def __init__(
        self, base_url: str, project: str, username: str | None, password: str | None
    ):
        self.base_url = base_url.rstrip("/")
        self.project = project
        self.session = requests.Session()
        if username and password:
            self.session.auth = (username, password)

    def _url(self, path: str) -> str:
        return f"{self.base_url}{path}"

    def list_repositories(self) -> list[str]:
        # GET /api/v2.0/projects/{project_name}/repositories
        url = self._url(f"/api/v2.0/projects/{self.project}/repositories")
        repos: list[str] = []
        page = 1
        page_size = 100

        while True:
            resp = self.session.get(
                url, params={"page": page, "page_size": page_size}, timeout=15
            )
            resp.raise_for_status()
            data = resp.json()
            if not isinstance(data, list):
                raise RuntimeError(
                    f"Unexpected Harbor response for repositories: {type(data)}"
                )

            for item in data:
                name = item.get("name")
                if isinstance(name, str):
                    # Harbor returns '<project>/<repo>'
                    repos.append(name)

            if len(data) < page_size:
                break
            page += 1

        return repos

    def list_artifacts(
        self, repository: str, *, with_tag: bool = True
    ) -> list[dict[str, Any]]:
        # GET /api/v2.0/projects/{project}/repositories/{repository_name}/artifacts
        # repository_name is the repository without the '<project>/' prefix.
        repo_short = repository
        prefix = f"{self.project}/"
        if repo_short.startswith(prefix):
            repo_short = repo_short[len(prefix) :]

        url = self._url(
            f"/api/v2.0/projects/{self.project}/repositories/{repo_short}/artifacts"
        )

        artifacts: list[dict[str, Any]] = []
        page = 1
        page_size = 100

        while True:
            resp = self.session.get(
                url,
                params={
                    "page": page,
                    "page_size": page_size,
                    "with_tag": str(with_tag).lower(),
                },
                timeout=15,
            )
            resp.raise_for_status()
            data = resp.json()
            if not isinstance(data, list):
                raise RuntimeError(
                    f"Unexpected Harbor response for artifacts: {type(data)}"
                )

            for item in data:
                if isinstance(item, dict):
                    artifacts.append(item)

            if len(data) < page_size:
                break
            page += 1

        return artifacts

    def list_latest_apps(self) -> list[HarborRepo]:
        """Return one entry per repo for tag 'latest' if present."""
        apps: list[HarborRepo] = []
        for repo in self.list_repositories():
            try:
                artifacts = self.list_artifacts(repo, with_tag=True)
            except requests.HTTPError:
                # Skip repos we cannot read
                continue

            for art in artifacts:
                tags = art.get("tags")
                if not isinstance(tags, list):
                    continue
                if any(isinstance(t, dict) and t.get("name") == "latest" for t in tags):
                    apps.append(HarborRepo(repo=repo, tag="latest"))
                    break

        apps.sort(key=lambda a: a.repo)
        return apps

File: rust-backend/src/harbor.rs
use anyhow::Context;
use base64::Engine;
use reqwest::StatusCode;
use serde::Deserialize;

use crate::models::HarborRepo;

#[derive(Clone)]
pub struct HarborClient {
    base_url: String,
    project: String,
    client: reqwest::Client,
}

impl HarborClient {
    pub fn new(
        base_url: String,
        project: String,
        username: Option<String>,
        password: Option<String>,
    ) -> anyhow::Result<Self> {
        let mut headers = reqwest::header::HeaderMap::new();
        headers.insert(
            reqwest::header::ACCEPT,
            reqwest::header::HeaderValue::from_static("application/json"),
        );

        if let (Some(u), Some(p)) = (username.as_ref(), password.as_ref()) {
            let token = base64::engine::general_purpose::STANDARD.encode(format!("{u}:{p}"));
            let v = reqwest::header::HeaderValue::from_str(&format!("Basic {token}"))?;
            headers.insert(reqwest::header::AUTHORIZATION, v);
        }

        let client = reqwest::Client::builder()
            .default_headers(headers)
            .build()?;

        Ok(Self {
            base_url: base_url.trim_end_matches('/').to_string(),
            project,
            client,
        })
    }

    fn url(&self, path: &str) -> String {
        format!("{}{}", self.base_url, path)
    }

    pub async fn list_latest_apps(&self) -> anyhow::Result<Vec<HarborRepo>> {
        let repos = self.list_repositories().await?;
        let mut apps = Vec::new();

        for repo in repos {
            match self.list_artifacts(&repo).await {
                Ok(arts) => {
                    if let Some(app) = resolve_latest_from_artifacts(&repo, &arts) {
                        apps.push(app);
                    }
                }
                Err(_) => {
                    continue;
                }
            }
        }

        apps.sort_by(|a, b| a.repo.cmp(&b.repo));
        Ok(apps)
    }

    async fn list_repositories(&self) -> anyhow::Result<Vec<String>> {
        #[derive(Debug, Deserialize)]
        struct RepoItem {
            name: Option<String>,
        }

        let mut repos = Vec::new();
        let mut page = 1;
        let page_size = 100;

        loop {
            let url = self.url(&format!("/api/v2.0/projects/{}/repositories", self.project));
            let resp = self
                .client
                .get(url)
                .query(&[("page", page), ("page_size", page_size)])
                .timeout(std::time::Duration::from_secs(15))
                .send()
                .await?;

            if !resp.status().is_success() {
                anyhow::bail!("Harbor repositories request failed: {}", resp.status());
            }

            let data: Vec<RepoItem> = resp.json().await?;
            for item in &data {
                if let Some(name) = &item.name {
                    repos.push(name.clone());
                }
            }

            if data.len() < page_size as usize {
                break;
            }
            page += 1;
        }

        Ok(repos)
    }

    async fn list_artifacts(&self, repository: &str) -> anyhow::Result<Vec<Artifact>> {
        let repo_short = repository
            .strip_prefix(&format!("{}/", self.project))
            .unwrap_or(repository);

        let mut artifacts = Vec::new();
        let mut page = 1;
        let page_size = 100;

        loop {
            let url = self.url(&format!(
                "/api/v2.0/projects/{}/repositories/{}/artifacts",
                self.project, repo_short
            ));

            let resp = self
                .client
                .get(url)
                .query(&[
                    ("page", page.to_string()),
                    ("page_size", page_size.to_string()),
                    ("with_tag", "true".to_string()),
                ])
                .timeout(std::time::Duration::from_secs(15))
                .send()
                .await?;

            if resp.status() == StatusCode::FORBIDDEN || resp.status() == StatusCode::UNAUTHORIZED {
                anyhow::bail!("Harbor artifacts request forbidden");
            }
            if !resp.status().is_success() {
                anyhow::bail!("Harbor artifacts request failed: {}", resp.status());
            }

            let data: Vec<Artifact> = resp.json().await.context("decode artifacts")?;
            artifacts.extend(data.iter().cloned());

            if data.len() < page_size as usize {
                break;
            }
            page += 1;
        }

        Ok(artifacts)
    }
}

#[derive(Clone, Debug, Deserialize)]
struct Artifact {
    tags: Option<Vec<Tag>>,
}

#[derive(Clone, Debug, Deserialize)]
struct Tag {
    name: Option<String>,
}

fn resolve_latest_from_artifacts(repo: &str, artifacts: &[Artifact]) -> Option<HarborRepo> {
    for art in artifacts {
        let tags = art.tags.as_ref()?;
        let mut tag_names: Vec<String> = tags.iter().filter_map(|t| t.name.clone()).collect();

        if !tag_names.iter().any(|t| t == "latest") {
            continue;
        }

        let resolved = resolve_tag(&tag_names);
        tag_names.sort();

        return Some(HarborRepo {
            repo: repo.to_string(),
            tag: resolved,
            all_tags: tag_names,
        });
    }

    None
}

pub fn resolve_tag(tag_names: &[String]) -> String {
    let numeric = tag_names
        .iter()
        .find(|t| t.chars().any(|c| c.is_ascii_digit()));
    if let Some(t) = numeric {
        return t.clone();
    }

    let non_latest = tag_names.iter().find(|t| t.as_str() != "latest");
    if let Some(t) = non_latest {
        return t.clone();
    }

    "latest".to_string()
}
